docs/EmbeddedSwift/UserManual.md

```diff
@@ -0,0 +1,138 @@
+
+```bash
+$ xcrun swiftc hello.swift -enable-experimental-feature Embedded -wmo -Osize -Xlinker -dead_strip
+$ nm -um ./hello
+                 (undefined) external _putchar (from libSystem)
+$ size -m ./hello
+Segment __TEXT: 16384
+  Section __text: 172
+...
+```
+
+## Embedded Swift is a subset of Swift
+
+Embedded Swift is a subset of the Swift language, and some features are not available in Embedded Swift, however features are available, including: Generics, protocols, enums with associated values, tuples, optionals, classes (instances are allocated on the heap and refcounted just like in regular Swift), inheritance, runtime polymorphism, arrays (heap-allocated copy-on-write just like in regular Swift) and many more.
+
+Features that are not available:
+
+- **Not available**: Runtime reflection (`Mirror` APIs).
+- **Not available**: Values of protocol types ("existentials"), e.g. `let a: Hashable = ...`, are not allowed. `Any` and `AnyObject` are also not allowed.
+- **Not available**: Metatypes, e.g. `let t = SomeClass.Type` or `type(of: value)` are not allowed.
+- **Not available yet (under development)**: The print() function for types other than StaticString and integers.
+- **Not available yet (under development)**: String. (StaticString **is** available).
+- **Not available yet (under development)**: Set.
+- **Not available yet (under development)**: Dictionary.
+
+## Libraries and modules in Embedded Swift
+
+TODO
+
+## Allocating and non-allocating Embedded Swift mode
+
+TODO
+
+## Runtime support for allocating Embedded Swift
+
+TODO
+
+## External dependencies
+
+Embedded Swift minimizes external dependencies, but they still exist. In Embedded Swift compilation mode, the compiler only triggers external dependencies based on actual usage of the program under compilation. The following table lists which situations cause which external dependencies:
+
+|                                                                     | external dependencies:                                                  |
+|---------------------------------------------------------------------|-------------------------------------------------------------------------|
+| *(basic dependencies, added by the compiler in various situations)* | `void *memset(void *, int, size_t);`                                    |
+| instantiating a class, or using UnsafePointer.allocate()            | `int posix_memalign(void **, size_t, size_t);`<br/>`void free(void *);` |
+| using print()                                                       | `int putchar(int);`                                                     |
+| using Hashable, Set, Dictionary, or random-number generating APIs   | `void arc4random_buf(void *, size_t);`                                  |
```

Comment:
There are also atomic intrinsics that should be accounted for reference types (if used).

Plus we should likely add a section about the requirement expectations of these functions; e.g. that putchar actually takes a uint8? or that arc4random_buf MUST be cryptographically secure etc.

Reply:
Re. putchar taking uint8 -- that's an oversight on my side, we need to match the standard putchar. Fixing that here: https://github.com/apple/swift/pull/70685/